flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala

```diff
@@ -627,24 +640,27 @@ abstract class TableEnvironment(val config: TableConfig) {
     * @param functionName name of the map function. Must not be unique but has to be a
     *                     valid Java class identifier.
     */
-  protected def sinkConversion[T](
-      physicalRowTypeInfo: TypeInformation[Row],
+  protected def sinkConversion[T, P](
```

Comment:
I would refactor this as follows:
- make `sinkConversion()` an abstract method that is implemented by `BatchTableEnvironemnt` and `StreamTableEnvironment`
- extract the code-generation part that converts a `Row` into the requested data type and keep this in a method `TableEnvironment.generateRowConverterFunction()`
- the implementations of `sinkConversion()` can use the `generateRowConverterFunction`